File: lachancuoicgoi_flutter/lib/ui/history_page/history_item_card.dart
import 'dart:convert';

import 'package:flutter/material.dart';

import '../../analysis/analysis_result.dart';
import '../../core/analysis_availability.dart';
import '../../core/risk_level.dart';
import '../../data/call_history.dart';
import '../theme/risk_level_colors.dart';

/// History item card matching Kotlin HistoryItemCard.kt
class HistoryItemCard extends StatelessWidget {
  const HistoryItemCard({super.key, required this.item, required this.onTap});

  final CallHistory item;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final cs = Theme.of(context).colorScheme;
    final tt = Theme.of(context).textTheme;
    final riskLevel = RiskLevel.fromString(item.riskLevel);
    final availability = AnalysisAvailability.fromStoredSession(
      recordingError: item.recordingError,
      hasTranscript: item.transcript.trim().isNotEmpty,
      analysisCompleted: _hasValidAnalysis(item.analysisResult),
    );
    final riskColor = availability.canShowRisk ? riskLevel.color : cs.outline;
    final statusLabel = _statusLabel(item.recordingErrorEnum, availability);
    // Keep the saved summary visible for legacy/corrupt analysis JSON. Risk is
    // still hidden when availability is insufficient, but replacing the
    // summary would discard the only useful content older records contain.
    final savedSummary = item.summary.trim();
    final summaryClaimsSafe =
        !availability.canShowRisk &&
        savedSummary.toLowerCase() ==
            RiskLevel.green.vietnameseName.toLowerCase();
    final displayedSummary = savedSummary.isNotEmpty && !summaryClaimsSafe
        ? item.summary
        : 'Không thể đưa ra kết luận từ dữ liệu hiện có.';

    return Card(
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
      color: cs.surfaceContainerHighest,
      clipBehavior: Clip.antiAlias,
      child: InkWell(
        onTap: onTap,
        child: IntrinsicHeight(
          child: Row(
            children: [
              // Risk color bar
              Container(width: 4, color: riskColor),

              // Content
              Expanded(
                child: Padding(
                  padding: const EdgeInsets.all(16),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      // Date + risk badge
                      Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: [
                          Text(item.dateTime, style: tt.bodySmall),
                          Container(
                            padding: const EdgeInsets.symmetric(
                              horizontal: 8,
                              vertical: 4,
                            ),
                            decoration: BoxDecoration(
                              color: riskColor.withValues(alpha: 0.1),
                              borderRadius: BorderRadius.circular(50),
                            ),
                            child: Text(
                              availability.canShowRisk
                                  ? riskLevel.vietnameseName
                                  : statusLabel,
                              style: TextStyle(
                                color: riskColor,
                                fontSize: 12,
                                fontWeight: FontWeight.bold,
                              ),
                            ),
                          ),
                        ],
                      ),
                      const SizedBox(height: 4),

                      if (!availability.canShowRisk) ...[
                        Row(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Icon(
                              Icons.info_outline,
                              size: 16,
                              color: cs.onSurfaceVariant,
                            ),
                            const SizedBox(width: 6),
                            Expanded(
                              child: Text(
                                availability.guidance,
                                style: tt.bodySmall?.copyWith(
                                  color: cs.onSurfaceVariant,
                                ),
                              ),
                            ),
                          ],
                        ),
                        const SizedBox(height: 4),
                      ],

                      // Summary — wrap in Semantics so screen reader reads
                      // the full text even when visually truncated.
                      Semantics(
                        label: displayedSummary,
                        child: Text(
                          displayedSummary,
                          style: tt.bodyLarge?.copyWith(
                            fontWeight: FontWeight.bold,
                          ),
                          maxLines: 2,
                          overflow: TextOverflow.ellipsis,
                        ),
                      ),
                      const SizedBox(height: 4),

                      // Duration + flag count
                      Text(
                        '${item.duration} • ${item.flagCount} dấu hiệu',
                        style: tt.bodySmall?.copyWith(
                          color: cs.onSurfaceVariant,
                        ),
                      ),
                      const SizedBox(height: 4),

                      // Analysis type static label
                      Container(
                        margin: const EdgeInsets.only(top: 8),
                        padding: const EdgeInsets.symmetric(
                          horizontal: 10,
                          vertical: 4,
                        ),
                        decoration: BoxDecoration(
                          color: item.analysisType == null
                              ? cs.surfaceContainerHighest.withValues(
                                  alpha: 0.5,
                                )
                              : cs.secondaryContainer,
                          borderRadius: BorderRadius.circular(8),
                          border: Border.all(
                            color: cs.outlineVariant.withValues(alpha: 0.5),
                          ),
                        ),
                        child: Text(
                          item.analysisType ?? 'Không phân tích',
                          style: tt.labelSmall?.copyWith(
                            color: item.analysisType == null
                                ? cs.onSurfaceVariant.withValues(alpha: 0.7)
                                : cs.onSecondaryContainer,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  static bool _hasValidAnalysis(String? raw) {
    if (raw == null || raw.trim().isEmpty) return false;
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map || decoded['overallRiskLevel'] == null) return false;
      final result = AnalysisResult.fromJson(decoded.cast<String, Object?>());
      return !result.isError;
    } on Object {
      return false;
    }
  }

  static String _statusLabel(
    RecordingError? error,
    AnalysisAvailability availability,
  ) => switch (error) {
    RecordingError.killed => 'Phiên bị gián đoạn',
    RecordingError.partial => 'Kết quả chưa hoàn chỉnh',
    _ => availability.vietnameseName,
  };
}

File: lachancuocgoi/app/src/main/java/com/example/lachancuocgoi/ui/HistoryPage/ItemHistory/HistoryItemCard.kt
package com.example.lachancuocgoi.ui.HistoryPage.ItemHistory

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SuggestionChip
import androidx.compose.material3.SuggestionChipDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.lachancuocgoi.RiskLevel
import com.example.lachancuocgoi.data.CallHistory

@Composable
fun HistoryItemCard(item: CallHistory, onClick: () -> Unit) {
    val riskLevel = try {
        RiskLevel.valueOf(item.riskLevel)
    } catch (e: IllegalArgumentException) {
        RiskLevel.GREEN
    }

    val riskColor = when (riskLevel) {
        RiskLevel.GREEN -> Color(0xFF4CAF50)
        RiskLevel.YELLOW -> Color(0xFFFFEB3B)
        RiskLevel.ORANGE -> Color(0xFFFF9800)
        RiskLevel.RED -> Color(0xFFF44336)
    }

    Card(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick),
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant)
    ) {
        Row(modifier = Modifier.height(IntrinsicSize.Min)) {
            Box(
                modifier = Modifier
                    .width(4.dp)
                    .fillMaxHeight()
                    .background(riskColor)
            )
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(item.dateTime, style = MaterialTheme.typography.bodySmall)
                    Box(
                        modifier = Modifier
                            .clip(RoundedCornerShape(50))
                            .background(riskColor.copy(alpha = 0.1f))
                            .padding(horizontal = 8.dp, vertical = 4.dp)
                    ) {
                        Text(
                            text = riskLevel.vietnameseName,
                            color = riskColor,
                            fontSize = 12.sp,
                            fontWeight = FontWeight.Bold
                        )
                    }
                }
                Text(
                    item.summary,
                    style = MaterialTheme.typography.bodyLarge,
                    fontWeight = FontWeight.Bold,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis
                )
                Text(
                    text = "${item.duration} • ${item.flagCount} dấu hiệu",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Spacer(modifier = Modifier.height(4.dp))
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.Start
                ) {
                    val analysisLabel = item.analysisType ?: "Không phân tích"

                    SuggestionChip(
                        onClick = { /* Do nothing for now */ },
                        label = { Text(analysisLabel) },
                        colors = if (item.analysisType == null) {
                            SuggestionChipDefaults.suggestionChipColors(containerColor = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.5f))
                        } else {
                            SuggestionChipDefaults.suggestionChipColors()
                        }
                    )
                }
            }
        }
    }
}